File: MtxVec.Delphi/YuleLevinson.pas
unit YuleLevinson;

interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Basic1, TeeProcs, TeEngine, Chart, StdCtrls, ExtCtrls, MtxVec,            
  Toeplitz, Series, ComCtrls,Math387,MtxVecEdit, Buttons, MtxExpr,         
  AbstractMtxVec;

type
  TYuleLev = class(TBasicForm1)
    Series1: TFastLineSeries;
    Series2: TFastLineSeries;
    Panel3: TPanel;
    RadioGroup1: TRadioGroup;
    TrackBar1: TTrackBar;
    Label1: TLabel;                           
    Label2: TLabel;
    Label3: TLabel;                                                             
    Label4: TLabel;
    TrackBar2: TTrackBar;                                  
    Label5: TLabel;
    Label6: TLabel;
    Button1: TBitBtn;
    procedure FormCreate(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure TrackBar1Change(Sender: TObject);
    procedure TrackBar2Change(Sender: TObject);
    procedure RadioGroup1Click(Sender: TObject);
  private
    y,spec,corr: Vector;
    CorrLen : Integer;
    ZeroPadding : Integer;
    LPCCoef : Integer;
    { Private declarations }
  public
    { Public declarations }
  end;

var
  YuleLev: TYuleLev;

implementation

Uses MtxVecTee, MtxVecInt;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

procedure TYuleLev.FormCreate(Sender: TObject);
begin
  inherited;
  With RichEdit1.Lines do
  begin
    Clear;
    Add('YuleWalker autoregressive spectra uses Levinson '
      + 'Durbin recursion to solve a toeplitz systems of '
      + 'linear equations taking only O(n2) operations '
      + 'instead of O(n3) as required by LUSolve. The chart '
      + 'compares FFT and YuleWalker AR. The corrLen defines '
      + 'the number of samples on which the Autocorrelation '
      + 'is performed and LPCCoef defines the number of '
      + 'computed autocorrelation coefficients. The method '
      + 'uses biased autocorrelation. FFT uses only LPCoef '
      + 'parameter to determine the number of sample to '
      + 'include. It then rounds LPCCoef to the nearest '
      + 'power of two. FFT uses no windowing.');
    Add('Zoom in on a chart (left-click and drag mouse '
      + 'over the chart) to see differences. Please note '
      + 'that it takes less then 10ms to compute a 32000 '
      + 'point FFT on P366.');
  end;
  RadioGroup1Click(RadioGroup1);
  TrackBar1Change(TrackBar1);
  TrackBar2Change(TrackBar2);
end;

procedure TYuleLev.Button1Click(Sender: TObject);
begin
  Series1.Clear;
  Series2.Clear;
  Screen.Cursor := crHourGlass;
  StartTimer;
  { Levinson Yule Walker }
  y.LoadFromFile('FFTData.vec'); {Load signal}
  y.Resize(CorrLen);
  corr := AutoCorrBiased(y,LPCCoef);  {auto correlation}
  Levinson(corr, y);            {Levinson recursion}
  y.Resize(LargestExp2(y.Length*ZeroPadding),True); {zero padder}
  spec := Log10(1/Mag(FFT(y)));
  Label5.Caption := 'Time needed for Levinson YW : '+ FormatSample(StopTimer*1000,'0.000') +' ms';

  DrawValues(spec,Series1,0,1,DownSize);

  StartTimer;
  { "Regular" FFT }
  y.LoadFromFile('FFTData.vec'); {Load signal}
  y.Resize(LPCCoef);
  y.Resize(LargestExp2(LPCCoef)*ZeroPadding,True);
  spec := Mag(FFT(y));
  spec := Log10(ThreshBottom(spec,1E-5));
  TimeElapsed := GetTickCount - TimeCheck;
  Label6.Caption := 'Time needed for FFT : '+ FormatSample(StopTimer*1000,'0.000')+' ms';
  DrawValues(spec,Series2,0,1,DownSize);

  Chart1.Repaint;
  Screen.Cursor := crDefault;
end;

procedure TYuleLev.TrackBar1Change(Sender: TObject);
begin
  CorrLen := TTrackBar(Sender).Position;
  Label2.Caption := IntToStr(CorrLen);
  TrackBar2.Max := CorrLen;
  TrackBar2.Frequency := CorrLen div 20;
  TrackBar2.Position := CorrLen div 2;
  TrackBar2Change(TrackBar2);
end;

procedure TYuleLev.TrackBar2Change(Sender: TObject);
begin
  LPCCoef := TTrackBar(Sender).Position;
  Label4.Caption := IntToStr(LPCCoef);
end;

procedure TYuleLev.RadioGroup1Click(Sender: TObject);
begin
  case TRadioGroup(Sender).ItemIndex of
  0    : ZeroPadding := 1;
  1    : ZeroPadding := 2;
  2    : ZeroPadding := 4;
  3    : ZeroPadding := 8;
  4    : ZeroPadding := 16;
  end;
end;      

initialization
   RegisterClass(TYuleLev);

end.
  

File: MtxVec.CPP/YuleLevinson.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "YuleLevinson.h"
#include "MtxVecTee.hpp"
#include "Toeplitz.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic1"
#pragma resource "*.dfm"
TYuleLev *YuleLev;
//---------------------------------------------------------------------------
__fastcall TYuleLev::TYuleLev(TComponent* Owner)
  : TBasicForm1(Owner)
{
  RadioGroup1Click(RadioGroup1);
  TrackBar1Change(TrackBar1);
  TrackBar2Change(TrackBar2);

  AnsiString str;

  RichEdit1->Lines->Clear();
  str =   "YuleWalker autoregressive spectra uses Levinson ";
  str +=  "Durbin recursion to solve a toeplitz systems of ";
  str +=  "linear equations taking only O(n2) operations ";
  str +=  "instead of O(n3) as required by LUSolve. The chart ";
  str +=  "compares FFT and YuleWalker AR. The corrLen defines ";
  str +=  "the number of samples on which the Autocorrelation ";
  str +=  "is performed and LPCCoef defines the number of ";
  str +=  "computed autocorrelation coefficients. The method ";
  str +=  "uses biased autocorrelation. FFT uses only LPCoef ";
  str +=  "parameter to determine the number of sample to ";
  str +=  "include. It then rounds LPCCoef to the nearest ";
  str +=  "power of two. FFT uses no windowing.";

  RichEdit1->Lines->Add(str);
  str =   "Zoom in on a chart (left-click and drag mouse ";
  str +=  "over the chart) to see differences. Please note ";
  str +=  "that it takes less then 10ms to compute a 32000 ";
  str +=  "point FFT on P366.";
  RichEdit1->Lines->Add(str);
}
//---------------------------------------------------------------------------


void __fastcall TYuleLev::Button1Click(TObject *Sender)
{
        sVector DownY;

        double Step;
		Series1->Clear();
        Series2->Clear();
        Screen->Cursor = crHourGlass;
        StartTimer();
        // Levinson Yule Walker
		y.LoadFromFile("FFTData.vec"); // Load signal
		y.Resize(CorrLen,false);
        corr.AutoCorrBiased(y,LPCCoef);  // auto correlation
        Levinson(corr, y); // Levinson recursion
		y.Resize(LargestExp2(y.Length*ZeroPadding),true); // zero padder
		x.FFT(y,false);
		spec.Mag(x);
		spec.Inv();
        spec.Log10();
        TimeElapsed = StopTimer()*1000.0;
        Label5->Caption = "Time needed for Levinson YW : "+ IntToStr((int) TimeElapsed)+" ms";
        if (DownSize)
		{
			DownY.PixelDownSample(Chart1->ChartWidth,spec,NULL,NULL);
			Step = 0.5*(spec.Length/double(Chart1->ChartWidth));
            DrawValues(DownY,Series1,0.0,Step,false);
        }
        else DrawValues(spec,Series1,0.0,1,false);

		StartTimer();
        // "Regular" FFT
		y.LoadFromFile("FFTData.vec"); // Load signal
		y.Resize(LPCCoef);
		y.Resize(LargestExp2(LPCCoef)*ZeroPadding,true);
		x.FFT(y,false);
		spec.Mag(x);
		spec.ThreshBottom(0.00001); //-100 dB
        spec.Log10();
        TimeElapsed = StopTimer()*1000;
        Label6->Caption = "Time needed for FFT : "+ IntToStr((int) TimeElapsed)+" ms";
        if (DownSize)
        {
			DownY.PixelDownSample(Chart1->ChartWidth,spec,NULL,NULL);
            Step = 0.5*(spec.Length/double(Chart1->ChartWidth));
            DrawValues(DownY,Series2,0.0,Step,false);
        }
        else DrawValues(spec,Series2,0.0,1,false);

        Chart1->Repaint();
        Screen->Cursor = crDefault;

}
//---------------------------------------------------------------------------

void __fastcall TYuleLev::TrackBar1Change(TObject *Sender)
{
  CorrLen = TrackBar1->Position;
  Label2->Caption = IntToStr(CorrLen);
  TrackBar2->Max = CorrLen;
  TrackBar2->Frequency = CorrLen*0.05;
  TrackBar2->Position = CorrLen*0.5;
  TrackBar2Change(TrackBar2);
}
//---------------------------------------------------------------------------

void __fastcall TYuleLev::TrackBar2Change(TObject *Sender)
{
  LPCCoef = TrackBar2->Position;
  Label4->Caption = IntToStr(LPCCoef);
}
//---------------------------------------------------------------------------

void __fastcall TYuleLev::RadioGroup1Click(TObject *Sender)
{
  switch (RadioGroup1->ItemIndex)
  {
    case 0 : ZeroPadding = 1; break;
    case 1 : ZeroPadding = 2; break;
    case 2 : ZeroPadding = 4; break;
    case 3 : ZeroPadding = 8; break;
    case 4 : ZeroPadding = 16; break;
  }
}
//---------------------------------------------------------------------------


 